explorer/interpreter/interpreter.cpp

```diff
@@ -860,6 +865,42 @@ auto Interpreter::Convert(Nonnull<const Value*> value,
       }
       return Convert(value, destination_type, source_loc);
     }
+    case Value::Kind::PointerValue: {
+      if (destination_type->kind() != Value::Kind::PointerType ||
+          cast<PointerType>(destination_type)->type().kind() !=
+              Value::Kind::NominalClassType) {
+        // No conversion needed.
+        return value;
+      }
+
+      // Get pointed value.
+      const auto* src_ptr = cast<PointerValue>(value);
+      CARBON_ASSIGN_OR_RETURN(const auto* pointee,
+                              heap_.Read(src_ptr->address(), source_loc))
+      if (pointee->kind() == Value::Kind::NominalClassType) {
+        // TODO: When does that happen? Just return as-is.
+        return value;
+      }
+
+      const auto* dest_ptr = cast<PointerType>(destination_type);
+      std::optional<Nonnull<const NominalClassValue*>> class_subobj =
+          cast<NominalClassValue>(pointee);
+      auto new_addr = src_ptr->address();
+      while (class_subobj) {
+        if (TypeEqual(&class_subobj.value()->type(), &dest_ptr->type(),
```

Comment:
This seems like too much "working out the semantics" to be doing in the interpreter -- I think the type-checker should instead work out how many levels of `.base` are implied by the implicit conversion, and add that information to the AST. Perhaps we should rewrite an implicit derived-to-base conversion into a series of `.base` member accesses?

Reply:
To try to figure out what this might look like -- are you suggesting something like adding rewrite support for all expressions, and inserting "OperatorExpression(As)" for each incremental step?

e.g., var `x: D* = A.MakePointer()` becomes `x: D* = A.MakePointer() as B* as C* as D*`, adding recursive rewrites onto the expression node that represents the initializer?

Note, I don't think I have a strong opinion either way -- we could state that implicit casts only work for one step, or that implicit casts work for the full reach. I'm not sure how much moving the logic to the type checker would be cleaner versus just hiding that this is still doing something similar (since in either case, this code assumes the type checker has already figured out the indicated cast is feasible).

Reply:
As a concrete example of how I'm imagining we might model this:

- Change `BuiltinConvertExpr` to be rewritable, following the same pattern as the other expression forms that derive from `RewritableMixin<T>`.
- When the type-checker performs an implicit conversion from derived to base, traversing N levels of class hierarchy, it rewrites that implicit conversion into `operand.base.base.base` where there are N levels of `.base`. Either:
    - Each level is represented as a `SimpleMemberAccessExpression` whose `member` is a `BaseElement`. This would mean that the `member` of a `SimpleMemberAccessExpression` changes from being a `NamedElement` to being any `Element`. Or,
    - Each level is represented by some new form of expression that performs a member access of a `BaseElement.

Note that we should already be building a `BuiltinConvertExpr` for every implicit conversion that uses one of the built-in rules, and we already build a call to `ImplicitAs.Convert` for every implicit conversion that uses a user-defined `impl`, so all the contexts in which an implicit conversion can happen already support having their operands rewritten; that wouldn't be new here.

Reply:
Thank you both.

I agree with Jon that I don't see it making a lot of difference today, because the fact that we try to convert here, means that there _is_ a valid subtyping conversion. I can rewrite the surrounding code to highlight that with `CARBON_FATAL` or `CHECK`.

But it would make more sense from a design standpoint for the interpreter to already have that information available yes. So same as Jon, no strong opinion.

I started looking at refactoring based on the example provided, thanks a lot!

Reply:
I'm happy for this to be changed to do more of the work in the type-checker and less in the interpreter in a follow-up pull request. And I'm going to be on vacation for a couple of weeks; so please don't block the progress of this patch on me or this comment thread!

Reply:
I think it'd make sense to merge this one first then. The suggested change is in progress in parallel, and that'd unblock the #2442 as well. Would that work for you @jonmeow ?

Reply:
I'm good with that plan; reviewed, approved and merging.